File: java/util/Either.java
package util;

public sealed abstract class Either<X, Y> {
    public abstract boolean isLeft();
    public abstract X left();
    public abstract Y right();

    public static <X, Y> Either<X, Y> ofLeft(final X value) {
        return new Left<>(value);
    }
    public static <X, Y> Either<X, Y> ofRight(final Y value) {
        return new Right<>(value);
    }

    private final static class Left<X, Y> extends Either<X, Y> {
        final X value;
        public Left(final X value) {
            this.value = value;
        }
        @Override
        public boolean isLeft() {
            return true;
        }
        @Override
        public X left() {
            return value;
        }
        @Override
        public Y right() {
            throw new RuntimeException("Trying to get right when left value is stored");
        }
    }

    private final static class Right<X, Y> extends Either<X, Y> {
        final Y value;

        public Right(final Y value) {
            this.value = value;
        }
        @Override
        public boolean isLeft() {
            return false;
        }
        @Override
        public X left() {
            throw new RuntimeException("Trying to get left when right value is stored");
        }
        @Override
        public Y right() {
            return value;
        }
    }
}

File: kotlin/util/Either.kt
package util

sealed class Either<out A, out B> {
    private class Left<A>(val value: A) : Either<A, Nothing>() {
        override val left: A
            get() = value
        override val right: Nothing
            get() = throw RuntimeException("Trying to get right when left value is stored")
        override val isLeft: Boolean
            get() = true
    }

    private class Right<B>(val value: B) : Either<Nothing, B>() {
        override val left: Nothing
            get() = throw RuntimeException("Trying to get left when right value is stored")
        override val right: B
            get() = value
        override val isLeft: Boolean
            get() = false
    }

    abstract val left: A
    abstract val right: B
    abstract val isLeft: Boolean

    companion object {
        fun <A, B> ofLeft(value: A): Either<A, B> {
            return Left(value)
        }

        fun <A, B> ofRight(value: B): Either<A, B> {
            return Right(value)
        }
    }
}